platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/service/WritableRosterStore.java

```diff
@@ -0,0 +1,233 @@
+    private final RosterStateAccessor rosterStateAccessor;
+    /**
+     * The roster state singleton. This is the state that holds the candidate roster hash and the list of pairs of
+     * active roster hashes and the round number in which those rosters became active.
+     *
+     * @implNote the use of {@link ReadablePlatformStateStore} and {@link WritablePlatformStateStore} to provide access
+     * to the roster states (beyond just the {@link PlatformState}) is deliberate, for convenience.
+     */
+    private final WritableSingletonState<RosterState> rosterState;
+
+    private final WritableKVState<ProtoBytes, Roster> rosterMap;
+
+    /**
+     * Constructs a new {@link WritableRosterStore} instance.
+     *
+     * @param writableStates the readable states
+     */
+    public WritableRosterStore(@NonNull final WritableStates writableStates) {
+        this.writableStates = writableStates;
+        this.rosterStateAccessor = new ReadableRosterStore(writableStates);
+        this.rosterState = writableStates.getSingleton(RosterStateId.ROSTER_STATES_KEY);
+        this.rosterMap = writableStates.get(RosterStateId.ROSTER_KEY);
+    }
+
+    /**
+     * Set the candidate roster if valid and doesn't yet exist in the state.
+     *
+     * @param candidateRoster a candidate roster to set.
+     */
+    @Override
+    public void setCandidateRoster(@NonNull final Roster candidateRoster) {
+        Objects.requireNonNull(candidateRoster);
+        RosterValidator.validate(candidateRoster);
+
+        // update the roster state
+        final RosterState currentRosterState = rosterStateOrThrow();
+        final Bytes candidateRosterHash = RosterUtils.hash(candidateRoster).getBytes();
+        final Builder rosterStateBuilder = RosterState.newBuilder()
+                .candidateRosterHash(candidateRosterHash)
+                .roundRosterPairs(currentRosterState.roundRosterPairs());
+        this.rosterState.put(rosterStateBuilder.build());
+
+        final Bytes currentCandidateRosterHash = currentRosterState.candidateRosterHash();
+        // remove existing candidate roster if present
+        this.rosterMap.remove(
+                ProtoBytes.newBuilder().value(currentCandidateRosterHash).build());
+
+        this.rosterMap.put(ProtoBytes.newBuilder().value(candidateRosterHash).build(), candidateRoster);
+        commit();
+    }
+
+    /**
+     * Determines the initial active roster based on the given software version and initial state.
+     * The active roster is obtained by adopting the candidate roster if a software upgrade is detected.
+     * Otherwise, the active roster is retrieved from the state or an exception is thrown if not found.
+     *
+     * @param version the software version of the current node
+     * @param initialState the initial state of the platform
+     * @return the active roster which will be used by the platform
+     */
+    @NonNull
+    @Override
+    public Roster determineActiveRoster(
```

Comment:
https://github.com/hashgraph/hedera-services/pull/15442#issuecomment-2388998839

Reply:
I'm still convinced we'll discover the utility of a feature flag here.  But since you're not invoking the method from a production code path, it's not needed in this PR.

Reply:
I'm reading the description of ticket https://github.com/hashgraph/hedera-services/issues/14712.   When I read this ticket, the ideas in my head include invoking the code in ServicesMain and gating the capability behind a feature flag while waiting for the rest of development to complete.